libkbfs/block_tree.go

```diff
@@ -269,179 +275,235 @@ func (bt *blockTree) getNextDirtyBlockAtOffset(ctx context.Context,
-// the block tree to leaf ("direct") blocks that encompass the given
-// offset range (half-inclusive) in the data.  If `endOff` is nil, it
-// returns blocks until reaching the end of the data.  If `prefixOk`
-// is true, the function will ignore context deadline errors and
-// return whatever prefix of the data it could fetch within the
-// deadine.  Return params:
-//
-//   * pathsFromRoot is a slice, ordered by offset, of paths from
-//     the root to each block that makes up the range.  If the path is
-//     empty, it indicates that pblock is a direct block and has no
-//     children.
-//   * blocks: a map from block pointer to a data-containing leaf node
-//     in the given range of offsets, if `getDirect` is true.
-//   * nextBlockOff is the offset of the block that follows the last
-//     block given in `pathsFromRoot`.  If `pathsFromRoot` contains
-//     the last block among the children, nextBlockOff is nil.
-func (bt *blockTree) getBlocksForOffsetRange(ctx context.Context,
-	ptr BlockPointer, pblock BlockWithPtrs, startOff, endOff Offset,
-	prefixOk bool, getDirect bool) (pathsFromRoot [][]parentBlockAndChildIndex,
-	blocks map[BlockPointer]Block, nextBlockOffset Offset,
-	err error) {
-	nextBlockOffset = pblock.FirstOffset()
-	if !pblock.IsIndirect() {
-		// Return a single empty path, under the assumption that the
-		// caller already checked the range for this block.
-		if getDirect {
-			// Return a child map with only this block in it.
-			return [][]parentBlockAndChildIndex{nil},
-				map[BlockPointer]Block{ptr: pblock}, nil, nil
-		}
-		// Return an empty child map with no blocks in it (since
-		// getDirect is false).
-		return [][]parentBlockAndChildIndex{nil}, nil, nil, nil
+// getBlocksForOffsetRangeTask is used for passing data to
+// getBlocksForOffsetRange tasks
+type getBlocksForOffsetRangeTask struct {
+	ptr        BlockPointer
+	pblock     BlockWithPtrs
+	pathPrefix []parentBlockAndChildIndex
+	startOff   Offset
+	endOff     Offset
+	prefixOk   bool
+	getDirect  bool
+}
+
+func (task *getBlocksForOffsetRangeTask) subTask(childPtr BlockPointer,
+	childPath []parentBlockAndChildIndex) getBlocksForOffsetRangeTask {
+	subTask := *task
+	subTask.ptr = childPtr
+	subTask.pblock = nil
+	subTask.pathPrefix = childPath
+	return subTask
+}
+
+// getBlocksForOffsetRangeResult is used for passing data back from
+// getBlocksForOffsetRange tasks
+type getBlocksForOffsetRangeResult struct {
+	pathFromRoot    []parentBlockAndChildIndex
+	ptr             BlockPointer
+	block           Block
+	nextBlockOffset Offset
+	err             error
+}
+
+// processGetBlocksTask examines the block it is passed, enqueueing any children
+// in range into wp, and passing data back through results.
+func (bt *blockTree) processGetBlocksTask(ctx context.Context,
+	wg *sync.WaitGroup, wp *workerpool.WorkerPool,
+	job getBlocksForOffsetRangeTask,
+	results chan getBlocksForOffsetRangeResult) {
+	defer wg.Done()
+
```

Comment:
Maybe the first thing we should do here is check that `ctx` hasn't been cancelled yet?